reaper_test.go

```diff
@@ -0,0 +1,726 @@
+		require.Contains(t, err.Error(), "image list: "+tc.imageListErr.Error())
+
+		require.Contains(t, log, "level=ERROR")
+		require.NotContains(t, log, "level=WARN")
+		require.Contains(t, log, "removed containers=2 networks=2 volumes=2 images=0")
+	})
+}
+
+// safeBuffer is a buffer safe for concurrent use.
+type safeBuffer struct {
+	buf bytes.Buffer
+	mtx sync.Mutex
+}
+
+// Write writes to the buffer.
+func (sb *safeBuffer) Write(p []byte) (n int, err error) {
+	sb.mtx.Lock()
+	defer sb.mtx.Unlock()
+
+	return sb.buf.Write(p)
+}
+
+// String returns the buffer as a string.
+func (sb *safeBuffer) String() string {
+	sb.mtx.Lock()
+	defer sb.mtx.Unlock()
+
+	return sb.buf.String()
+}
+
+func TestAbortedClient(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	t.Cleanup(cancel)
+
+	var log safeBuffer
+	logger := withLogger(slog.New(slog.NewTextHandler(&log, &slog.HandlerOptions{
+		Level: slog.LevelDebug,
+	})))
+	tc := newRunTest()
+	client := newMockClient(tc)
+	r, err := newReaper(ctx, logger, withClient(client), testConfig)
+	require.NoError(t, err)
+
+	// Start processing clients.
+	go r.processClients()
+
+	// Fake a shutdown to trigger the client abort.
+	close(r.shutdown)
+
+	addr := r.listener.Addr().String()
+
+	var d net.Dialer
+	conn, err := d.DialContext(ctx, "tcp", addr)
+	require.NoError(t, err)
+
+	// With shutdown triggered the client should be aborted.
+	// The write will still succeed due to buffering but the read will fail.
+	_, err = conn.Write([]byte("some-filter\n"))
+	require.NoError(t, err)
+
+	buf := make([]byte, 4)
+	n, err := conn.Read(buf)
+	require.Error(t, err)
+	switch {
+	case errors.Is(err, io.EOF),
+		errors.Is(err, syscall.ECONNRESET):
+		// Expected errors.
+	default:
+		t.Fatal("unexpected read error:", err)
+	}
+	require.Zero(t, n)
+	require.Contains(t, log.String(), "shutdown, aborting client")
+}
+
+func TestShutdownTimeout(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	t.Cleanup(cancel)
+
+	var log safeBuffer
+	logger := withLogger(slog.New(slog.NewTextHandler(&log, &slog.HandlerOptions{
+		Level: slog.LevelDebug,
+	})))
+	tc := newRunTest()
+	client := newMockClient(tc)
+	r, err := newReaper(ctx, logger, withClient(client), testConfig)
+	require.NoError(t, err)
+
+	errCh := make(chan error, 1)
+	runCtx, runCancel := context.WithCancel(ctx)
+	t.Cleanup(runCancel)
+	go func() {
+		errCh <- r.run(runCtx)
+	}()
+
+	require.NoError(t, err)
+
+	testConnect(ctx, t, r.listener.Addr().String())
+	runCancel()
+
+	select {
+	case err = <-errCh:
+		require.NoError(t, err)
+	case <-ctx.Done():
+		t.Fatal("timeout", log.String())
+	}
+
+	require.Contains(t, log.String(), "shutdown timeout")
+}
+
+func TestReapContainer(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	t.Cleanup(cancel)
+
+	// Run a test container.
+	args := []string{"container", "run", "--detach", "--quiet"}
```

Comment:
What do you prefer: 1) shelling out to the binary, 2) using the docker client from the docker/docker dependency (we already have it), or 3) keeping testcontainers-go as dogfooding, showcasing the benefits of using it?

Reply:
I wanted to avoid using testcontainers-go as that creates a circular repository dependency, which has bitten me in the past.

Good call on client though, simpler to use except for the image, which isn't too bad.

Reply:
He says and then ContainerCreate doesn't pull missing images :(